### ISSUE TITLE
Next paper candidates: 29 Nov

### ISSUE THREAD

**Issue by m09** [src-d/reading-club#94]
# Next paper candidates
Let's propose papers to study next! All papers mentioned in the comments of this issue will be listed in the next vote.

**Comment by m09** [comment]
Last session runner-up: [Deep Learning Anti-patterns from Code Metrics History](https://arxiv.org/abs/1910.07658)

> Anti-patterns are poor solutions to recurring design problems. Number of empirical studies have highlighted the negative impact of anti-patterns on software maintenance which motivated the development of various detection techniques. Most of these approaches rely on structural metrics of software systems to identify affected components while others exploit historical information by analyzing co-changes occurring between code components. By relying solely on one aspect of software systems (i.e., structural or historical), existing approaches miss some precious information which limits their performances.
> In this paper, we propose CAME (Convolutional Analysis of code Metrics Evolution), a deep-learning based approach that relies on both structural and historical information to detect anti-patterns. Our approach exploits historical values of structural code metrics mined from version control systems and uses a Convolutional Neural Network classifier to infer the presence of anti-patterns from this information. We experiment our approach for the widely known God Class anti-pattern and evaluate its performances on three software systems. With the results of our study, we show that: (1) using historical values of source code metrics allows to increase the precision; (2) CAME outperforms existing static machine-learning classifiers; and (3) CAME outperforms existing detection tools.

**Comment by bzz** [comment]
[Learning to Fix Build Errors with Graph2Diff Neural Networks](http://www.cs.toronto.edu/~dtarlow/papers/graph2diff_preprint.pdf), Google

> Professional software developers spend a significant amount of
time fixing builds, but this has received little attention as a problem in automatic program repair. We present a new deep learning
architecture, called Graph2Diff, for automatically localizing and
fixing build errors. We represent source code, build configuration
files, and compiler diagnostic messages as a graph, and then use a
Graph Neural Network model to predict a diff. A diff specifies how
to modify the code’s abstract syntax tree, represented in the neural
network as a sequence of tokens and of pointers to code locations.
Our network is an instance of a more general abstraction which we
call Graph2Tocopo, which is potentially useful in any development
tool for predicting source code changes. We evaluate the model on
a dataset of over 500k real build errors and their resolutions from
professional developers. Compared to the approach of DeepDelta
[23], our approach tackles the harder task of predicting a more
precise diff but still achieves over double the accuracy.

**Comment by sara-02** [comment]
What about the latest one by Github-Microsoft, [CodeSearchNet Challenge: Evaluating the State of Semantic Code Search](https://arxiv.org/abs/1909.09436)

> Semantic code search is the task of retrieving relevant code given a natural language query. While related to other information retrieval tasks, it requires bridging the gap between the language used in code (often abbreviated and highly technical) and natural language more suitable to describe vague concepts and ideas.
To enable evaluation of progress on code search, we are releasing the CodeSearchNet Corpus and are presenting the CodeSearchNet Challenge, which consists of 99 natural language queries with about 4k expert relevance annotations of likely results from CodeSearchNet Corpus. The corpus contains about 6 million functions from open-source code spanning six programming languages (Go, Java, JavaScript, PHP, Python, and Ruby). The CodeSearchNet Corpus also contains automatically generated query-like natural language for 2 million functions, obtained from mechanically scraping and preprocessing associated function documentation. In this article, we describe the methodology used to obtain the corpus and expert labels, as well as a number of simple baseline solutions for the task.
We hope that CodeSearchNet Challenge encourages researchers and practitioners to study this interesting task further and will host a competition and leaderboard to track the progress on the challenge. We are also keen on extending CodeSearchNet Challenge to more queries and programming languages in the future.

**Comment by bzz** [comment]
[Adversarial Examples for Models of Code](https://arxiv.org/abs/1910.07517), code2vec
> We introduce a novel approach for attacking trained models of code with adversarial examples. The main idea is to force a given trained model to make a prediction of the adversary's choice by introducing small perturbations that do not change program semantics. We find these perturbations by deriving the desired prediction with respect to the model's inputs while holding the model weights constant and following the gradients to slightly modify the input. 
To defend a model against such attacks, we propose placing a defensive model in front of the downstream model. The defensive model detects unlikely mutations and masks them before feeding the input to the downstream model. 
We show that our attack succeeds in changing a prediction to the adversary's desire ("targeted attack") up to 89% of the times, and succeeds in changing a given prediction to any incorrect prediction ("non-targeted attack") 94% of the times. By using our proposed defense, the success rate of the attack drops drastically for both targeted and non-targeted attacks, with a minor penalty of 2% relative degradation in accuracy while not performing under attack.

**Comment by sara-02** [comment]
Hey @m09 @bzz will there be a session this Friday(i.e 29th)?

**Comment by bzz** [comment]
> Hey @m09 @bzz will there be a session this Friday(i.e 29th)?

Hey, sorry for delay - yes! Things were quite hectic last couple of weeks. We are going to do the vote ASAP now

**Comment by m09** [comment]
[Poll](https://forum.sourced.tech/t/2019-11-29-paper-tbd/272/2) is up!

**Comment by m09** [comment]
Results are out, we'll study [CodeSearchNet Challenge: Evaluating the State of Semantic Code Search](https://drive.google.com/open?id=1VtuMU-M1MkwShRbGSzlUn7BSumCqn7O9) :smile_cat:
